### PR TITLE
Allow for overriding of .csdp.cache file for micromega

### PR DESCRIPTION
**Pull request by Alizter**
We check for a new env var MICROMEGA_CACHE_FILE which if set to
"cache_file" will read the file "cache_file.cache", copy it to a new
file "cache_file.cache.new" and continue using the new file as the
persistent csdp cache, as .csdp.cache was being used before.

If the env var is not set, the old behaviour of reading and writing to
.csdp.cache is kept.

This will allow us to manage the micromega cache in a controlled way
which will be needed for porting the test-suite to dune.

The issue with .csdp.cache in the dune test-suite is that .csdp.cache is a dependency for many dune rules whilst at the same time being mutable. This makes it very tricky to write dune rules for, and can often invalidate cache / reproducibility. For this reason, dune forbids writeable targets and dependencies. By allowing the specification of the cache file we can have more control in the test-suite over who modifies what.

This should not change anything for anybody. Though I would like to hear what @coq/micromega-maintainers think about doing things this way.

cc @ejgallego 

- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.
  - [ ] Documented any new / changed **user messages**.
  - [ ] Updated **documented syntax** by running `make -f Makefile.dune doc_gram_rsts`.
- [ ] Opened **overlay** pull requests.

### CODE REVIEW

**Comment by Alizter** [comment]
OK so it looks like that wasn't the best place to plop that down, since it is performance critical. I suppose some sort of global state is in order.

**Comment by ejgallego** [comment]
Thanks @Alizter , we need to understand the security implications of this as we discussed privately. Likely restricting the cache ability to cross directories should suffice, but we need to check best practices here.

**Comment by Alizter** [comment]
I've opted for a single MICROMEGA_CACHE_FILE which reads from a file with a .cache extension and creates a .cache.new file. Even if directories are added, this should be safer than what we had before.

File: plugins/micromega/coq_micromega.ml [hunk 1]
```diff
@@ -64,6 +64,37 @@ let use_nra_cache =
...
+  (* Default behavior *)
+  | None -> ".csdp.cache"
+  (* When the enviornment variable "MICROMEGA_CACHE_FILE" is set to
+  "cache_file", we copy the file "cache_file.cache" to "cache_file.cache.new"
+  and then use "cache_file.cache.new" as the persistant cache (what .csdp.cache
+  was being used for). *)
+  | Some cache_file ->
+    let cache_source = cache_file ^ ".cache" in
+    let cache_target = cache_file ^ ".cache.new" in
+    (* Copy binary file *)
```
Comment (fajb) [inline plugins/micromega/coq_micromega.ml:79]: Wouldn't there be a better home for a `copy` function ? Maybe lib/system.ml.

**Comment by fajb** [comment]
Somehow it seems overkill to add an environment variable for the purpose of building the test-suite.
An alternative would be to rebuild the `.csdp.cache` just like other caches. This would *just* require having `csdp` installed.
Would that be too costly? 

Anyhow, if you are willing to have this env variable, that's fine for me.

**Comment by Alizter** [comment]
After some thought I've decided that we perhaps don't want to go down this route so I will close this PR.

**Comment by ejgallego** [comment]
@Alizter I am not sure the setup you mentioned is sound w.r.t. build semantics, so we may in the end to need this.

**Comment by Alizter** [comment]
@ejgallego Let's discuss on Zulip then.

**Pull request closed without merging**